src/compose/application-manager.ts

```diff
@@ -234,6 +236,7 @@ export async function getRequiredSteps(
 
 		// We want to remove images before moving on to anything else
 		if (steps.length === 0) {
+			// TODO: this should be comparing by uuid if available
```

Comment:
If we're completely supporting UUIDs, the keys for the apps object in the target state from v3 target state endpoint would be UUID, so we could step in [here](https://github.com/balena-os/balena-supervisor/blob/master/src/compose/application-manager.ts#L200-L201) and check the key type with regex, like so:
```
const appIdRegex = /^[0-9]+$/;
const currentAppIdsOrUuids = Object.keys(currentApps).map((i) => i.match(appIdRegex) ? parseInt(i, 10) : i);
const targetAppIdsOrUuids = Object.keys(targetApps).map((i) => i.match(appIdRegex) ? parseInt(i, 10) : i);
```
This change would satisfy the TODO comment you made here.

Reply:
Or just check key length, if UUID length is consistent and appIds will never reach the same length as UUIDs. Regex feels safer though

Reply:
Update: targetApps parameter is always [keyed by appId](https://github.com/balena-os/balena-supervisor/blob/master/src/device-state/db-format.ts#L31-L34), so this change is unnecessary.